source/common/tls/connection_info_impl_base.cc

```diff
@@ -26,6 +36,16 @@ ConnectionInfoImplBase::getCachedValueOrCreate(CachedValueTag tag,
     const ValueType* val = absl::get_if<ValueType>(&it->second);
     ASSERT(val != nullptr, "Incorrect type in variant");
     if (val != nullptr) {
+
+      // Some values are retrieved too early, for example if properties of a peer certificate are
+      // retrieved before the handshake is complete, an empty value is cached. The value must be
+      // in the cache, so that we can return a valid reference, but in those cases if another caller
+      // later retrieves the same value, we must recalculate the value.
+      if (shouldRecalculateCachedEntry(*val)) {
+        it->second = create(ssl());
+        val = &absl::get<ValueType>(it->second);
```

Comment:
I think this logic doesn't put non-empty value back into cache? That means if the cache was queries too early, the cache is not effective.

Reply:
`it->second = create(ssl());` is putting it into the cache; it's being stored into the same hashmap node value (without changing the key). I validated that this does get stored (and looked up and used) for subsequent accesses.